nixos/modules/tasks/auto-upgrade.nix

```diff
@@ -158,113 +180,185 @@ in {
-      '';
+          [
+            "--refresh"
+            "--flake ${cfg.flake}"
+          ]
+      );
+    }
+
+    (lib.mkIf cfg.enable {
+      systemd.services.nixos-upgrade = {
+        description = "NixOS Upgrade";
+
+        restartIfChanged = false;
+        unitConfig.X-StopOnRemoval = false;
+
+        serviceConfig.Type = "oneshot";
+
+        environment =
+          config.nix.envVars
+          // {
+            inherit (config.environment.sessionVariables) NIX_PATH;
+            HOME = "/root";
+          }
+          // config.networking.proxy.envVars;
+
+        path = with pkgs; [
+          coreutils
+          gnutar
+          xz.bin
+          gzip
+          gitMinimal
+          config.nix.package.out
+          config.programs.ssh.package
+        ];
 
-      startAt = cfg.dates;
+        script =
+          let
+            nixos-rebuild = "${config.system.build.nixos-rebuild}/bin/nixos-rebuild";
+            date = "${pkgs.coreutils}/bin/date";
+            readlink = "${pkgs.coreutils}/bin/readlink";
+            shutdown = "${config.systemd.package}/bin/shutdown";
+            upgradeFlag = lib.optional (cfg.channel == null) "--upgrade";
+          in
+          if cfg.allowReboot then
+            ''
+              ${nixos-rebuild} boot ${toString (cfg.flags ++ upgradeFlag)}
+              booted="$(${readlink} /run/booted-system/{initrd,kernel,kernel-modules})"
+              built="$(${readlink} /nix/var/nix/profiles/system/{initrd,kernel,kernel-modules})"
+
+              ${lib.optionalString (cfg.rebootWindow != null) ''
+                current_time="$(${date} +%H:%M)"
+
+                lower="${cfg.rebootWindow.lower}"
+                upper="${cfg.rebootWindow.upper}"
+
+                if [[ "''${lower}" < "''${upper}" ]]; then
+                  if [[ "''${current_time}" > "''${lower}" ]] && \
+                     [[ "''${current_time}" < "''${upper}" ]]; then
+                    do_reboot="true"
+                  else
+                    do_reboot="false"
+                  fi
+                else
+                  # lower > upper, so we are crossing midnight (e.g. lower=23h, upper=6h)
+                  # we want to reboot if cur > 23h or cur < 6h
+                  if [[ "''${current_time}" < "''${upper}" ]] || \
+                     [[ "''${current_time}" > "''${lower}" ]]; then
+                    do_reboot="true"
+                  else
+                    do_reboot="false"
+                  fi
+                fi
+              ''}
+
+              if [ "''${booted}" = "''${built}" ]; then
+                ${nixos-rebuild} ${cfg.operation} ${toString cfg.flags}
+              ${lib.optionalString (cfg.rebootWindow != null) ''
+                elif [ "''${do_reboot}" != true ]; then
+                  echo "Outside of configured reboot window, skipping."
+              ''}
+              else
+                ${shutdown} -r +1
+              fi
```

Comment:
Would you be OK to pull this out to a separate .bash file? It would need to be passed (or `--replace-fail`ed) a bunch (which is unfortunate), but it would be easier to review and lint/format with tools like `shfmt` and `shellcheck`. Not a show-stopper, though.